File: cinenotes-backend/src/main/java/com/cinenotes/domain/Review.java
package com.cinenotes.domain;

import java.time.LocalDateTime;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.cinenotes.user.AppUser;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(
        name = "reviews",
        uniqueConstraints = @UniqueConstraint(
                name = "uk_reviews_user_title",
                columnNames = {"user_id", "title_id"}
        )
)
@Getter
@Setter
@NoArgsConstructor
public class Review {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private AppUser user;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "title_id", nullable = false)
    private Title title;

    private Double rating;

    @Column(name = "review_text", columnDefinition = "TEXT")
    private String reviewText;

    @Column(name = "review_language")
    private String language;

    @Column(name = "contains_spoiler", nullable = false)
    private Boolean containsSpoiler = false;

    @Column(name = "is_visible", nullable = false)
    private Boolean isVisible = true;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
}

File: cinenotes-frontend/src/types/review.ts
import type { UserSummaryResponse } from "./user";

export type ReviewCreateRequest = {
  titleId: number;
  rating: number;
  reviewText: string | null;
  reviewLanguage: string | null;
  containsSpoiler: boolean | null;
};

export type ReviewUpdateRequest = {
  rating: number | null;
  reviewText: string | null;
  reviewLanguage: string | null;
  containsSpoiler: boolean | null;
};

export type ReviewResponse = {
  id: number;
  user: UserSummaryResponse;
  titleId: number;
  titleName: string;
  rating: number;
  reviewText: string | null;
  reviewLanguage: string | null;
  containsSpoiler: boolean | null;
  visible: boolean | null;
  createdAt: string;
  updatedAt: string;
};
